### ISSUE TITLE
not working

### ISSUE THREAD

**Issue by boistordu** [fjudith/docker-draw.io#3]
first your readme is wrong since we need to had draw in the url because if we do not we go to a tomcat test page.

Second if I go to draw/ I have a ssl problem and not the same one as the previous issue. 
`ERR_SSL_PROTOCOL_ERROR`
Is it because I'm not using it in localhost but from a distant point in the LAN? Does your container only work in loop?

**Comment by boistordu** [comment]
docker logs shows:
```
11-Sep-2017 07:07:18.777 INFO [main] org.apache.catalina.startup.Catalina.start Server startup in 937 ms
11-Sep-2017 07:27:36.633 INFO [http-nio-8080-exec-7] org.apache.coyote.http11.Http11Processor.service Error parsing HTTP request header
 Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.
 java.lang.IllegalArgumentException: Invalid character found in method name. HTTP method names must be tokens
	at org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:422)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:683)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1457)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)


```

**Comment by fjudith** [comment]
Hello @boistordu 

The following code in the Dockerfile put `draw` webapp at the root `/`.

```
cd $CATALINA_HOME && \
    xmlstarlet ed \
    -P -S -L \
    -i '/Server/Service/Engine/Host/Valve' -t 'elem' -n 'Context' \
    -i '/Server/Service/Engine/Host/Context' -t 'attr' -n 'path' -v '/' \
    -i '/Server/Service/Engine/Host/Context[@path="/"]' -t 'attr' -n 'docBase' -v 'draw' \
    -s '/Server/Service/Engine/Host/Context[@path="/"]' -t 'elem' -n 'WatchedResource' -v 'WEB-INF/web.xml' \
    -i '/Server/Service/Engine/Host/Valve' -t 'elem' -n 'Context' \
    -i '/Server/Service/Engine/Host/Context[not(@path="/")]' -t 'attr' -n 'path' -v '/ROOT' \
    -s '/Server/Service/Engine/Host/Context[@path="/ROOT"]' -t 'attr' -n 'docBase' -v 'ROOT' \
    -s '/Server/Service/Engine/Host/Context[@path="/ROOT"]' -t 'elem' -n 'WatchedResource' -v 'WEB-INF/web.xml' \
conf/server.xml
```
Maybe you're using an old image.

The `draw` webapps redirects the url to https.
As a workaround i suggest you put an haproxy ssl-offloader in front of `draw.io`.
Or use the `https=0` in the url.

**example**
```
http://192.168.99.100:32761/?https=0 
```

I personnally run this image in production using HAProxy.

Note that i'm planning to implement a patch to disable URL rewrite before compiling the webapp.

**Comment by boistordu** [comment]
Well it's new image so we can already exclude that problem. Or your latest doesn't apply to the REAL latest.

Anyway you can do the test yourself, http://ip: port/ you'll see the tomcat page so I don't know what's wrong with the code but apparently something is. I can do a video if you want, just to show that's not in effect your @root position.

For the proxy problem, I was not planning to use any proxy and I find that a bit overkill no? Why the proxy isn't combined with it if it s a requirement?

**Comment by boistordu** [comment]
So let's show it to you :
`3f13f46be544        fjudith/draw.io               "catalina.sh run"        14 hours ago        Up 14 hours              0.0.0.0:50102->8080/tcp                                                                                                                                                                                                                draw`
that was the launching command 
`sudo docker run -it --rm --name="draw" -p 50102:8080 -d fjudith/draw.io`
I've pulled your image this morning for me so 15 hours ago. Now I've repulled with latest specified and it seems that you've update the build? 
Anyway I've just tried again and it's still the same result. 

Isn't be a security issue to arrive at the tomcat webpage? (even if it's obviously not supposed to happen)  

So here are the result anyway: 
![pourgithub2](https://user-images.githubusercontent.com/9523075/30297160-1569b3c4-9747-11e7-8f5a-e195da8db620.png)
![pourgithub3](https://user-images.githubusercontent.com/9523075/30297161-157c1c8a-9747-11e7-80ae-fd27b3408589.png)
![pourgithubdraw](https://user-images.githubusercontent.com/9523075/30297159-156642e8-9747-11e7-8d43-b6f064fb3ad5.png)

but one positive thing, the update you've made have made gone the java error. 

Sorry about the proxy but beside privoxy I've never used one because never really needed, but if you point to me to the right documentation i can try to put it in place and see what happens then. 
But anyway draw is not at root position obviously and I don't feel safe about the tomcat page. 
I can give you access to the contianer to see by yourself if you want to?

**Comment by boistordu** [comment]
Sorry I missed a few lines on my mail. 
So yes apparently we need to restart the container to have tomcat put the thing at root position. 
So do you think there is a way to improve the behaviour of tomcat? 
And when you said using HAproxy, you mean another container linked to this one then ?

**Comment by boistordu** [comment]
And, something just strucked me... if it redirects to https, then it means that It want to use another port no? And if we can't use https it's because this other port is not open right? or is it totally something else? Because if it's just that, I can totally open this other port too...

**Comment by fjudith** [comment]
The first issue is solved.
Draw.IO displays on the first run `http://ip:port/http=0`

![image](https://user-images.githubusercontent.com/17888379/30299986-935a5a6c-9752-11e7-914b-368ed2969c4c.png)

Now dealing with the certificate auto-gen using Let's Encrypt. But il will required several environment variables (e.g Common Name, Org, etc.)

**Comment by boistordu** [comment]
ooow so it's certificate problem. OKey I understand. do you have any specific folders, configuration file which has to be modified because of the certificate? 

did you do another rebuild after my message?

**Comment by boistordu** [comment]
apparently at first startup it show the good content. great.

**Comment by fjudith** [comment]
I simply unziped the webapps in order that Tomcat detects the `draw` subfolder and validate the redirection on the first run.
 
I'm currently managing to implement `SSL` listening on port 8443.
Their will be 3 supported scenarios
- Generate SELF signed certificate (default)
- Generate Let's encrypt certificate (e.g `LETS_ENCRYPT_ENABLED=true`)
- Do nothing if keystore mounted as an external keystore (e.g. `-v /home/user/kestore.jks:/usr/local/tomcat/.keystore)

This will take me some time as i should setup a Travis CI to automate the tests

**Comment by fjudith** [comment]
@boistordu 

SSL certificate is implemented on port 8443 in the latest image `7.3.6, latest`.
Documentation has been updated as well.

**Comment by boistordu** [comment]
it works well :) thanks .
